add initial CONTRIBUTING.md (#148)
kudobuilder/frameworks: CONTRIBUTING.md

# Contributing Guidelines

The KUDO Operators project accepts contributions via GitHub pull requests. This document outlines the process to help get your contribution accepted.

## Sign the CLA

> Kubernetes projects require that you sign a Contributor License Agreement (CLA) before we can accept your pull requests.

Reviewer: Why does that apply to KUDO?

Author: @mattj-io can elaborate more but this is part of the preparation we need to become a CNCF project

Reviewer 2: Because the whole project has been submitted as CNCF sandbox. We need to follow the same governance model.

Author: moved to DCO, as thats where KUDO is moving now
https://github.com/kudobuilder/kudo/pull/1164

Reviewer: There's more discussions needed here, we're not sure what applies to KUDO right now.

Please see https://git.k8s.io/community/CLA.md for more info.

## How to Contribute to an Existing Operator

> 1. Submit an issue describing your proposed change to the operator in question.

Reviewer 3: markdown nit: I think you can do 1. 1. 1. instead of 1. 2. 3. to make the future edits easier

Author: neat!

> 1. The operator maintainer or repo owners will respond  to your issue promptly

Reviewer 4: ```suggestion
1. The operator maintainer or repository owners will respond to your issue promptly
```

> 1. If your proposed change is accepted, and you haven't already done so, sign a Contributor License Agreement (see details above).

Reviewer 4: ```suggestion
1. If your proposed change is accepted and you haven't already done so, sign the Contributor License Agreement (see details above).
```

> 1. Fork the desired repo, develop and test your code changes.

Reviewer 4: ```suggestion
1. Fork the KUDO Operators repository, develop and test your code changes in the relevant Operator directory.
```

1. Submit a pull request.

> ### Requirements for the new Operators

Reviewer 4: ```suggestion
### Requirements for new Operators
```

- Must include a `README.md`

> - Must include source GitHub repositories for images used in the Operator

Reviewer 3: hmm what does this one actually mean? that I use images in docs and I need to link sources? 🤔

Author: let me fix that, I meant container images

> - Must include source GitHub repositories for the container images used in the Operator

Reviewer 4: ```suggestion
- Must include source GitHub repositories and/or Dockerfiles for Docker images used in the Operator
```

> - Must pass the `kubectl kudo package verify` 

Reviewer 4: ```suggestion
- Must pass `kubectl kudo package verify` 
```

- Must specify the minimum `KUDO` version and `Kubernetes` version in `operator.yaml`

> - Must provide tests for the Operator

Reviewer 3: let's make this more specific - maybe we start with requiring at least a single test that executed deploy plan?

> - Must provide at least one deploy plan test for the Operator

Reviewer 4: ```suggestion
- Must provide at least one passing deploy plan test for the Operator
```

- Must follow the structure described in [getting started guide](https://kudo.dev/docs/developing-operators.html#getting-started)
- Each file in `templates` directory must be of one Kubernetes object type
- Should follow Kubernetes best practices

> - Should support Operator upgrades

Reviewer 3: do we really require this? what if it's something simple that does not need a special plan for upgrades?

Author: I wanted to separate `must` and `should` and this I included in `should`.

> - Should follow Kubernetes best practices

Reviewer 4: Can we link said best practices?

Author: we can add, a subset of practices here which we consider relevant. As most links might include how to configure Kubernetes practices also

Author: actually, to avoid ambiguous and open to interpretation rules, I have removed this line as for now

- Should provide a sane default configuration

> ### Release Process

Reviewer 3: 👍 for being transparent here!

> Release process will be automated in upcoming iterations. Right now there are two ways to manually trigger the release process

Reviewer 4: ```suggestion
The release process will be automated in upcoming iterations. Right now there are two ways to manually trigger the release process:
```

- Submit an issue describing which operator has to be released

> - Use the [KUDO Slack](https://kubernetes.slack.com/messages/kudo/) channel to ask the owners of the repository to release an operator

Reviewer 4: ```suggestion
- Use the [KUDO Slack channel](https://kubernetes.slack.com/messages/kudo/) to ask the owners of the repository to release the operator
```
